File: tools/cpp/KML/tests/statistics/test_StreamingMean.cc
// unittest_streaming_means

#include <gtest/gtest.h>
#include "statistics/StreamingMean.h"
#include <vector>

using KML::Statistics::StreamingMean;

class StreamingMeanTest : public ::testing::Test
{
    public:
        StreamingMean* sm;

        StreamingMeanTest()
        {
            sm = new StreamingMean(2);
        }
};

// Evaluate with no updates should return 0.0 for mean.
TEST_F(StreamingMeanTest, TestEmptyEvaluate)
{
    EXPECT_FLOAT_EQ(0.0, sm->evaluate());
}

// Mean value should come out to 10 when all values are same.
TEST_F(StreamingMeanTest, TestMeanCalculation)
{
    std::vector<double> data{10, 10};
    sm->update(data);
    EXPECT_FLOAT_EQ(10.0, sm->evaluate());
}

// Mean value should come out to 10 when all values are same.
TEST_F(StreamingMeanTest, TestMeanDriftCalculation)
{
    std::vector<double> data{2, 1, 1};
    sm->update(data);
    EXPECT_FLOAT_EQ(1.0, sm->evaluate());
}

File: tools/python/KML/tests/statistics/test_StreamingMean.py
import unittest

from KML.statistics.StreamingMean import PyStreamingMean


class TestStreamingMean(unittest.TestCase):

    WINDOW_SIZE = 5
    DATA = [10.0] * WINDOW_SIZE
    SM = None

    @classmethod
    def setUpClass(cls):
        cls.SM = PyStreamingMean(cls.WINDOW_SIZE)

    @classmethod
    def tearDownClass(cls):
        cls.SM = None

    def test_update(self):
        self.SM.update(self.DATA)

    def test_evaluate(self):
        self.SM.update(self.DATA)
        self.assertEqual(self.SM.evaluate(), 10.0)

    def test_drift(self):
        data = [10, 10, 10, 10, 10, 5, 5, 5, 5, 5]
        self.SM.update(data)
        self.assertEqual(self.SM.evaluate(), 5.0)


if __name__ == "__main__":
    unittest.main()
